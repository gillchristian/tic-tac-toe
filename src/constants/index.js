export const O = '●'
export const X = '✖'
export const s = ''

export const RESULTS = [
  /1..1..1../,
  /.1..1..1./,
  /..1..1..1/,
  /111....../,
  /...111.../,
  /......111/,
  /1...1...1/,
  /..1.1.1../
]

