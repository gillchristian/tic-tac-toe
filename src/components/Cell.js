import R from 'ramda'

import React, { PropTypes } from 'react'
import styled from 'styled-components'

import { X, O } from '../constants'

const getCellColor = R.compose(
  R.cond([
    [R.equals(X), R.always('palevioletred')],
    [R.equals(O), R.always('tomato')],
    [R.T,         R.always('papayawhip')],
  ]),
  R.prop('mark')
)

const Wrapper = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;
  background-color: ${getCellColor};
  width: 64px;
  height: 64px;
  border-radius: 14px;
  cursor: pointer;
  margin: 5px;
  font-weight: bold;
  font-size: 40px;
  color: white;
  box-shadow: 0 2px 5px gray;
  user-select: none;

  &:hover {
    background-color: ${props => !props.mark && 'lightgray'};
    cursor: ${props => !!props.mark && 'default'};
`

class Cell extends React.Component {
  static propTypes = {
    mark: PropTypes.oneOf([O, X, '']),
    next: PropTypes.oneOf([O, X])
  }

  state = {
    mouseIsOver: false
  }

  mouseEnter = () => {
    this.setState({ mouseIsOver: true })
  }

  mouseLeave = () => {
    this.setState({ mouseIsOver: false })
  }

  render() {
    const {mark, next} = this.props

    return (
      <Wrapper
        mark={mark}
        next={next}
        onClick={this.props.onClick}
        onMouseEnter={this.mouseEnter}
        onMouseLeave={this.mouseLeave}
      >
        {
          mark
            ? mark
            : (this.state.mouseIsOver ? next : '')
        }
      </Wrapper>
    )
  }
}

export default Cell

