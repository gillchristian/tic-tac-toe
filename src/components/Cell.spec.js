import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'

import Cell from './Cell'

import { X, O, s } from '../constants'

describe('Cell', () => {
  it('renders ok', () => {
    const wrapper = shallow(<Cell next={X} mark={O} />)

    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('changes state when mouse enters and leaves', () => {
    const wrapper = shallow(<Cell next={X} mark={O} />)

    expect(wrapper.state()).toEqual({ mouseIsOver: false })

    wrapper.simulate('mouseenter')

    expect(wrapper.state()).toEqual({ mouseIsOver: true })

    wrapper.simulate('mouseleave')

    expect(wrapper.state()).toEqual({ mouseIsOver: false })
  })

  it('renders the mark if present, or the next mark on hover if present', () => {
    const wrapper = shallow(<Cell next={X} mark={O} />)

    expect(wrapper.children().text()).toEqual(O)

    wrapper.simulate('mouseenter')

    // props.mark is present, still displayes the mark
    expect(wrapper.children().text()).toEqual(O)

    wrapper.setProps({ next: X, mark: s })

    // mouse still over, no mark so displayes next mark
    expect(wrapper.children().text()).toEqual(X)

    wrapper.simulate('mouseleave')

    // mouse not over, no mark, displayes no mark
    expect(toJson(wrapper)).toMatchSnapshot()

    wrapper.setProps({ next: s, mark: s })
    wrapper.simulate('mouseenter')

    // mouse over, no mark, no next mark, displayes no mark
    expect(toJson(wrapper)).toMatchSnapshot()
  })
})

