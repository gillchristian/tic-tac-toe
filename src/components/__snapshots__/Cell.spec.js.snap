// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cell renders ok 1`] = `
<styled.div
  mark="O"
  next="X"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  O
</styled.div>
`;

exports[`Cell renders the mark if present, or the next mark on hover if present 1`] = `
<styled.div
  mark=""
  next="X"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
/>
`;

exports[`Cell renders the mark if present, or the next mark on hover if present 2`] = `
<styled.div
  mark=""
  next=""
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
/>
`;
